pkg/faketikv/client.go

```diff
@@ -25,5 +39,273 @@ type Client interface {
+		return nil
+	}
+
+	return errors.Trace(errFailInitClusterID)
+}
+
+func (c *client) getMembers(ctx context.Context) (*pdpb.GetMembersResponse, error) {
+	members, err := c.pdClient().GetMembers(ctx, &pdpb.GetMembersRequest{})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return members, nil
+}
+
+func (c *client) createGRPCConn() (*grpc.ClientConn, error) {
+	cc, err := grpc.Dial(strings.TrimLeft(c.url, "http://"), grpc.WithInsecure())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return cc, nil
+}
+
+func (c *client) createHeartbeatStream() (pdpb.PD_RegionHeartbeatClient, context.Context, context.CancelFunc) {
+	var (
+		stream pdpb.PD_RegionHeartbeatClient
+		err    error
+		cancel context.CancelFunc
+		ctx    context.Context
+	)
+	for {
+		ctx, cancel = context.WithCancel(c.ctx)
+		stream, err = c.pdClient().RegionHeartbeat(ctx)
+		if err != nil {
+			log.Errorf("[store %d][pd] create region heartbeat stream error: %v", c.tag, err)
+			cancel()
+			select {
+			case <-time.After(time.Second):
+				continue
+			case <-c.ctx.Done():
+				log.Info("cancel create stream loop")
+				return nil, ctx, cancel
+			}
+		}
+		break
+	}
+	return stream, ctx, cancel
+}
+
+func (c *client) heartbeatStreamLoop() {
+	defer c.wg.Done()
+	for {
+		stream, ctx, cancel := c.createHeartbeatStream()
+		if stream == nil {
+			return
+		}
+		errCh := make(chan error, 1)
+		wg := &sync.WaitGroup{}
+		wg.Add(2)
+		go c.reportRegionHeartbeat(ctx, stream, errCh, wg)
+		go c.receiveRegionHeartbeat(ctx, stream, errCh, wg)
+		select {
+		case err := <-errCh:
+			log.Infof("[store %d][pd] heartbeat stream get error: %s ", c.tag, err)
+			cancel()
+		case <-c.ctx.Done():
+			log.Info("cancel heartbeat stream loop")
+			return
+		}
+		wg.Wait()
+	}
+}
+
+func (c *client) receiveRegionHeartbeat(ctx context.Context, stream pdpb.PD_RegionHeartbeatClient, errCh chan error, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		resp, err := stream.Recv()
+		if err != nil {
+			errCh <- err
+			return
+		}
+		select {
+		case c.receiveRegionHeartbeatCh <- resp:
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *client) reportRegionHeartbeat(ctx context.Context, stream pdpb.PD_RegionHeartbeatClient, errCh chan error, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for {
+		select {
+		case region := <-c.reportRegionHeartbeatCh:
+			request := &pdpb.RegionHeartbeatRequest{
+				Header:          c.requestHeader(),
+				Region:          region.Region,
+				Leader:          region.Leader,
+				DownPeers:       region.DownPeers,
+				PendingPeers:    region.PendingPeers,
+				BytesWritten:    region.WrittenBytes,
+				BytesRead:       region.ReadBytes,
+				ApproximateSize: uint64(region.ApproximateSize),
+			}
+			err := stream.Send(request)
+			if err != nil {
+				errCh <- err
+				log.Errorf("[store %d][pd] report regionHeartbeat error: %v", c.tag, err)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (c *client) Close() {
+	c.cancel()
+	c.wg.Wait()
```

Comment:
do we need to close the connection before wait?

Reply:
both are ok.